### ISSUE TITLE
 Libc call mlock failed with result in GoDaddy.Asherah.SecureMemory

### ISSUE THREAD

**Issue by lyricnz** [godaddy/asherah#260]
**Describe the bug**
Attempting to encrypt content in a Lambda function, got following exception a number of times.  Slightly sanitised stack trace below.  The calling code is

``` return session.Encrypt(Encoding.UTF8.GetBytes(text)).ToString();```

where session is a fairly long-lived result of 
```public Session<byte[], JObject> GetSessionBytesAsJson(string partitionId)```
from https://github.com/godaddy/asherah/blob/master/csharp/AppEncryption/AppEncryption/SessionFactory.cs

```
[Error] OURAPP.Lambda.EditItemsLambda: EditItemsLambda failed for user APIUSER GoDaddy.Asherah.Crypto.Exceptions.AppEncryptionException: Failed call action method, error: Libc call mlock failed with result -1 12 ---> GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.Libc.LibcOperationFailedException: Libc call mlock failed with result -1 12
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.Libc.LibcProtectedMemoryAllocatorLP64.CheckZero(Int32 result, String methodName)
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.Libc.LibcProtectedMemoryAllocatorLP64.Alloc(UInt64 length)
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.ProtectedMemorySecret..ctor(Byte[] sourceBytes, IProtectedMemoryAllocator allocator)
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.ProtectedMemorySecretFactory.CreateSecret(Byte[] secretData)
   at GoDaddy.Asherah.Crypto.AeadCrypto.GenerateKeyFromBytes(Byte[] sourceBytes, DateTimeOffset created, Boolean revoked)
   at GoDaddy.Asherah.Crypto.AeadCrypto.GenerateRandomCryptoKey(DateTimeOffset created)
   at GoDaddy.Asherah.Crypto.AeadCrypto.GenerateRandomCryptoKey()
   at GoDaddy.Asherah.Crypto.Envelope.AeadEnvelopeCrypto.EnvelopeEncrypt(Byte[] plainText, CryptoKey keyEncryptionKey, Object userState)
   at GoDaddy.Asherah.AppEncryption.Envelope.EnvelopeEncryptionJsonImpl.ApplyFunctionAndDisposeKey[T](CryptoKey key, Func`2 functionWithKey)
   --- End of inner exception stack trace ---
   at GoDaddy.Asherah.AppEncryption.Envelope.EnvelopeEncryptionJsonImpl.ApplyFunctionAndDisposeKey[T](CryptoKey key, Func`2 functionWithKey)
   at GoDaddy.Asherah.AppEncryption.Envelope.EnvelopeEncryptionJsonImpl.WithIntermediateKeyForWrite[T](Func`2 functionWithIntermediateKey)
   at GoDaddy.Asherah.AppEncryption.Envelope.EnvelopeEncryptionJsonImpl.EncryptPayload(Byte[] payload)
   at OURAPP.SDK.Encryption.EncryptionService.Encrypt(String text) in ...SDK.Encryption/EncryptionService.cs:line 26
```

**To Reproduce**
Run lambda repeatedly, possibly in parallel.

**Expected behavior**
No exception, working. 

**Additional context**
Code was recently changed from creating the session for every encrypt/decrypt operation, to creating it once early on, and reusing it.   It's possible that it's being used by more than one Thread at a time, with background Tasks??   (I'm not a c# expert).  Do we need to lock() on session?

**Comment by nikoo28** [comment]
The issue is probably caused due to reaching lockable memory limits. Can you try the same operations with session caching enabled? This may help to reuse the memory rather than trying to allocate new memory for every session.

**Comment by lyricnz** [comment]
There would have only been one session in place above.  I had move the session-creation to startup(ish), and was reusing it each time.  Do we need to lock() on session if there is a chance of concurrent access?

**Comment by lyricnz** [comment]
I just got this again, even with long-lived sessions and lock() on the session before use:
```
            lock (session)
            {
                return session.Encrypt(Encoding.UTF8.GetBytes(text)).ToString();
            }
```
Stack trace looks the same as above.  Would session caching make any difference?

**Comment by lyricnz** [comment]
We're .NET core 2.1 on AWS Lambda, fwiw.

**Comment by lyricnz** [comment]
How can I get diagnostics about how many/much memory is used here? There should only be ~1 session here.

**Comment by lyricnz** [comment]
12 ENOMEM Cannot allocate memory.  The new process image required more memory than was allowed by the hardware or by system-imposed memory management constraints.  A lack of swap space is normally temporary; however, a lack of core is not.  Soft limits may be increased to their corresponding hard limits.

**Comment by lyricnz** [comment]
This is a critical issue for us - the library is unusable as it fails very often.

**Comment by lyricnz** [comment]
FWIW, exactly the same problem if all key+session caching is disabled, as suggested in godaddy-oss #ashera channel.

**Comment by sushantmimani** [comment]
Hi @lyricnz ,

Please see linked gist for a working code snippet that uses Asherah for encrypt and decrypt operations using AWS Lambda.
https://gist.github.com/sushantmimani/706c20c004bc5868177265c590ba1999

Can you check if you're doing something similar?

**Comment by lyricnz** [comment]
Yes, our code looks pretty similar. Just spend 3 hours working through this with @sushantmimani .  Can reproduce the crash with gist code above. 
```
2020-07-22T23:18:10.518000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 START RequestId: 24393697-dad8-4219-b8c0-c7a38ba8971a Version: $LATEST
2020-07-22T23:18:10.787000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 [Information] OURCOMPANY.OURAPP.Lambda.CreateStuff: CreateStuff called by organisation APITESTUSER
2020-07-22T23:18:10.797000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 [Debug] OURCOMPANY.OURAPP.Lambda.CreateStuff: Setting FeatureRoleArn from environment variable to arn:aws:iam::OURAWSACCOUNT:role/dr4fs-FeatureRole-simonroberts
2020-07-22T23:18:10.858000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 [Debug] OURCOMPANY.OURAPP.Lambda.CreateStuff: Retrieving tenant context for tenant ID APITESTUSER from table named Tenants-Tenants-dev
2020-07-22T23:18:10.859000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 [Information] OURCOMPANY.SDK.DynamoDB.Tenant.TenantRepository: Getting tenant with ID "APITESTUSER"
2020-07-22T23:18:10.939000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 [Information] OURCOMPANY.SDK.DynamoDB.DynamoDbContextProvider: Injecting policy: removed
2020-07-22T23:18:12.125000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 [Debug] OURCOMPANY.OURAPP.Lambda.CreateStuff: Tenant context for tenant ID APITESTUSER retrieved successfully
2020-07-22T23:18:12.137000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 About to run test function
2020-07-22T23:18:12.247000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC0
2020-07-22T23:18:12.722000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.726000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC1
2020-07-22T23:18:12.738000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.738000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC2
2020-07-22T23:18:12.745000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.745000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC3
2020-07-22T23:18:12.752000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.752000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC4
2020-07-22T23:18:12.759000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.759000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC5
2020-07-22T23:18:12.766000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.766000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC6
2020-07-22T23:18:12.773000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.773000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC7
2020-07-22T23:18:12.779000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.779000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC8
2020-07-22T23:18:12.785000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.785000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC9
2020-07-22T23:18:12.791000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.791000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC10
2020-07-22T23:18:12.797000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.797000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC11
2020-07-22T23:18:12.803000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.803000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC12
2020-07-22T23:18:12.809000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 matches = True
2020-07-22T23:18:12.809000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 partition is shopperABC13
2020-07-22T23:18:12.831000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 [Error] OURCOMPANY.OURAPP.Lambda.CreateStuff: CreateStuff failed for user APITESTUSER GoDaddy.Asherah.Crypto.Exceptions.AppEncryptionException: Failed call action method, error: Libc call mlock failed with result -1 12
 ---> GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.Libc.LibcOperationFailedException: Libc call mlock failed with result -1 12
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.Libc.LibcProtectedMemoryAllocatorLP64.CheckZero(Int32 result, String methodName)
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.Libc.LibcProtectedMemoryAllocatorLP64.Alloc(UInt64 length)
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.ProtectedMemorySecret..ctor(Byte[] sourceBytes, IProtectedMemoryAllocator allocator)
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.ProtectedMemorySecretFactory.CreateSecret(Byte[] secretData)
   at GoDaddy.Asherah.SecureMemory.TransientSecretFactory.CreateSecret(Byte[] secretData)
   at GoDaddy.Asherah.Crypto.AeadCrypto.GenerateKeyFromBytes(Byte[] sourceBytes, DateTimeOffset created, Boolean revoked)
   at GoDaddy.Asherah.Crypto.AeadCrypto.GenerateKeyFromBytes(Byte[] sourceBytes, DateTimeOffset created)
   at GoDaddy.Asherah.Crypto.AeadCrypto.GenerateRandomCryptoKey(DateTimeOffset created)
   at GoDaddy.Asherah.Crypto.AeadCrypto.GenerateRandomCryptoKey()
   at GoDaddy.Asherah.Crypto.AeadCrypto.GenerateKey()
   at GoDaddy.Asherah.Crypto.Envelope.AeadEnvelopeCrypto.EnvelopeEncrypt(Byte[] plainText, CryptoKey keyEncryptionKey, Object userState)
   at GoDaddy.Asherah.AppEncryption.Envelope.EnvelopeEncryptionJsonImpl.<>c__DisplayClass13_0.<EncryptPayload>b__0(CryptoKey intermediateCryptoKey)
   at GoDaddy.Asherah.AppEncryption.Envelope.EnvelopeEncryptionJsonImpl.ApplyFunctionAndDisposeKey[T](CryptoKey key, Func`2 functionWithKey)
   --- End of inner exception stack trace ---
   at GoDaddy.Asherah.AppEncryption.Envelope.EnvelopeEncryptionJsonImpl.ApplyFunctionAndDisposeKey[T](CryptoKey key, Func`2 functionWithKey)
   at GoDaddy.Asherah.AppEncryption.Envelope.EnvelopeEncryptionJsonImpl.WithIntermediateKeyForWrite[T](Func`2 functionWithIntermediateKey)
   at GoDaddy.Asherah.AppEncryption.Envelope.EnvelopeEncryptionJsonImpl.EncryptPayload(Byte[] payload)
   at GoDaddy.Asherah.AppEncryption.SessionFactory.CachedSession.EncryptPayload(Byte[] payload)
   at GoDaddy.Asherah.AppEncryption.Envelope.EnvelopeEncryptionBytesImpl.EncryptPayload(Byte[] payload)
   at GoDaddy.Asherah.AppEncryption.SessionBytesImpl`1.Encrypt(Byte[] payload)
   at OURCOMPANY.OURAPP.Lambda.CreateStuff.MyFunctionHandler(String input, ILambdaContext context) in /my/source/dir/OURCOMPANY/OURAPP/OURCOMPANY.OURAPP/OURCOMPANY.OURAPP.Lambda/CreateStuff.cs:line 63
   at OURCOMPANY.OURAPP.Lambda.CreateStuff.Main(APIGatewayProxyRequest request, ILambdaContext context, IServiceProvider serviceProvider) in /my/source/dir/OURCOMPANY/OURAPP/OURCOMPANY.OURAPP/OURCOMPANY.OURAPP.Lambda/CreateStuff.cs:line 90
   at OURCOMPANY.SDK.Lambda.LambdaApiGatewayHandler.Function(APIGatewayProxyRequest request, ILambdaContext context) in /my/source/dir/OURCOMPANY/OURAPP/OURCOMPANY.SDK/OURCOMPANY.SDK.Lambda/LambdaApiGatewayHandler.cs:line 87
2020-07-22T23:18:12.863000+00:00 2020/07/22/[$LATEST]37e8f5f3b9c44f8588b5b969f7aa9145 END RequestId: 24393697-dad8-4219-b8c0-c7a38ba8971a
```

**Comment by lyricnz** [comment]
The test failed about 20 runs in a row, each managing somewhere between 10-14 iterations of the loop before failing.
This "seemed" to happen mostly when session-caching was enabled, but our (other) tests have it fail the same way eventually anyway, so suspect it's a likelihood not a on/off thing.

**Comment by lyricnz** [comment]
OK, it looks like Asherah is attempting to prevent allocations that exceed the mlock limit

```
        public virtual IntPtr Alloc(ulong length)
        {
            libc.getrlimit(GetMemLockLimit(), out var rlim);
            if (rlim.rlim_max != rlimit.UNLIMITED && rlim.rlim_max < length)
            {
                throw new MemoryLimitException(
                    $"Requested MemLock length exceeds resource limit max of {rlim.rlim_max}");
            }
```

https://github.com/godaddy/asherah/blob/master/csharp/SecureMemory/SecureMemory/ProtectedMemoryImpl/Libc/LibcProtectedMemoryAllocatorLP64.cs#L51

But this is comparing the size of a **single** allocation against the **total** amount available to lock. And in fact comparing rlim_max rather than rlim_cur which could?? be lower.

Regardless, by attempting a stupidly big allocation using the following code

```
            byte[] secretBytes = new byte[100000000];
            ISecretFactory secretFactory = new ProtectedMemorySecretFactory();
            using (Secret secretKey = secretFactory.CreateSecret(secretBytes))
            {
                secretKey.WithSecretBytes(bytes =>
                {
                    Console.Write($"##### got some bytes {bytes.Length}");
                    // DoSomethingWithSecretBytes(bytes);
                });
            }
```

I was able to get an exception that described the limit (for me, on Lambda) of 64KB. 

```
2020-07-23T12:54:09.197+10:00
[Error] *OUTSTUFF* GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.MemoryLimitException: Requested MemLock length exceeds resource limit max of 65536
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.Libc.LibcProtectedMemoryAllocatorLP64.Alloc(UInt64 length)
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.ProtectedMemorySecret..ctor(Byte[] sourceBytes, IProtectedMemoryAllocator allocator)
   at GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.ProtectedMemorySecretFactory.CreateSecret(Byte[] secretData)
   at OURCODE
```

This seems a very low limit (and I'm surprised it's not zero, for a shared-execution environment!)

Q: what is the page-size on this platform?  
Q: how many allocations do we put in this secure memory?
Q: how can we ensure that they are munlock() as quickly as possible?  Relying on the garbage collector and finalizers will likely be too late when the heap is large relative to the amount of mlock available.

**Comment by lyricnz** [comment]
FWIW increasing the max heap to 3008 MB didn't change the rlim_max

**Comment by lyricnz** [comment]
OK, I think I've got to the bottom of it:

- in Lambda there is a very small amount of memory available for mlock() - 64KB, which makes it critical to release those locks as soon as possible

- In an environment with lots of heap, the GC (and subsequent finalizer/Dispose) might not happen for some time, meaning that we're holding onto it much longer than we should.

- This can be mitigated somewhat by implementing IDisposable on the classes that hold references to SessionFactory etc, and explicitly calling their Dispose()

- But we're still going to have a hard-failure when there are enough lumps of SecureMemory to fill up 64KB (in units of pagesize) during a single ... session

**Comment by lyricnz** [comment]
From http://www.lambdashell.com/

```
user@host:~ ulimit -a
core file size          (blocks, -c) unlimited
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 736
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 736
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

**Comment by lyricnz** [comment]
FWIW the pagesize is 4KB on x86 Linux, which means we can hold just 16 pieces of SecureMemory.

**Comment by lyricnz** [comment]
For example, if we turn on session caching, that's enough to push us over the limit again sometimes.

**Comment by sushantmimani** [comment]
Hey @lyricnz 
Asherah relies on the fact that enough lockable memory exists for the use case to work and we cannot remove the usage of mLock as that is essential to the security of the library. We've tried running the sample code multiple times, with session caching disabled and it did not run into any mlock issues at all. If you could provide us with the a collated gist/document of the following, it will help us debug this further:
- Asherah config used for your use case
- AWS env config parameters
- Sample code that shows Asherah integration

**Comment by lyricnz** [comment]
We can quite reliably get the mlock problem, if you turn on session caching with your code. This fails at ~12 iterations, after presumably there are ~16 total SecureMemory items.

**Comment by aka-bo** [comment]
@lyricnz
> We can quite reliably get the mlock problem, if you turn on session caching with your code. This fails at ~12 iterations, after presumably there are ~16 total SecureMemory items.

Modifying the [example function](https://github.com/godaddy/asherah/issues/260#issuecomment-662712111) to enable session caching doesn't expose a bug, it merely demonstrates expected behavior. The lockable memory requirement described in [System Requirements](https://github.com/godaddy/asherah/blob/master/docs/SystemRequirements.md) _must_ be accounted for in any implementation, AWS Lambda-based implementations included.

The sample provided by @sushantmimani is an example of one such "known good" implementation. It demonstrates the use of multiple sessions via a single session factory, works within the known environmental constraints, and should be adaptable to cover both of the use-cases described in the issue description:

> Code was recently changed from creating the session for every encrypt/decrypt operation, to creating it once early on, and reusing it.

The main caveat being of course, that the adapted implementation would also need to account for the known env contraints by, for example, restricting the total number of secure memory resources while also disposing of those resources explicitly via `using` blocks (as opposed to leaving it up to the GC).

This leads us back to @sushantmimani's request. If you have an example implementation that you believe accounts for these env constraints and continues to run into "mlock failed" error, please provide the example/docs and we'd be happy to assist.

**Comment by lyricnz** [comment]
System requirements say only that 4KB/key are required.

- Does that include DRK?
- What about temporary/transient use of SecureMemory?
- What memory/mlock grows within a session if it is used for more than a single operation?  (seemingly not connected to number of Keys). 
- What additional secure-memory is required by session cache? The simple case, as demonstrated by the code above, fails when session caching is enabled (even though there should realistically only be one session).

When there is only (laughable) 64KB available in AWS Lamda, these questions suddenly become important for us.

**Comment by sushantmimani** [comment]
Hi @lyricnz . Please find our responses to your questions below:

> System requirements say only that 4KB/key are required.
> 
> * Does that include DRK? - _Yes, it does._
> * What about temporary/transient use of SecureMemory? _SecureMemory is used every time a new key is created._
> * What memory/mlock grows within a session if it is used for more than a single operation?  (seemingly not connected to number of Keys). - _Usage of SecureMemory is directly connected to the number of keys._
> * What additional secure-memory is required by session cache? The simple case, as demonstrated by the code above, fails when session caching is enabled (even though there should realistically only be one session). - _No, the case above does not use a single session. Since every partition is unique, the code creates 100 new session (https://gist.github.com/sushantmimani/706c20c004bc5868177265c590ba1999#file-asherah-lambda-cs-L49) and if you enable session caching, you're effectively trying to save 100 different sessions which will not work with lambda.If the code is updated to run 10 iterations, then it completes successfully with session caching enabled._

Additionally, please note that, as per the SystemRequirements doc an entire page is locked for each key. 4KiB is just the typical page size. This _might_ vary based on the processor architecture.

**Comment by lyricnz** [comment]
Please add metrics or logging so we can tell what securememory is being allocated/freed and when.

Or better still, allow pluggable implementation of SecureMemory so library consumers can make choices about the allocator that works on their platform. I know 64KB is not enough for general use, but the protections offered by SecureMemory aren't super-applicable to a managed shared execution environment like AWS Lambda.  I'm positive other library users will need this in future, so it's not "just us".

**Comment by lyricnz** [comment]
This is still an issue for people running Asherah in AWS Lambda.

**Comment by lyricnz** [comment]
STILL an issue. Created feature request for pluggable secure-memory https://github.com/godaddy/asherah/issues/290

**Comment by jgowdy** [comment]
The solution to this that is in the pipeline is switching to an OpenSSL based secure heap implementation, with added protections on top.  The mmap implementation is inefficient with 4KB pages per secret/key.  On non-lambda systems like Kubernetes/EKS or traditional hosting, the mlock limits are configurable.

Once we get the OpenSSL based secure heap implementation done for Java, the memory usage will be drastically more efficient.

**Comment by jgowdy** [comment]
Work in progress to fix the C# implementation with regards to 4KB heap allocation and mlock limits:

Phase 1: https://github.com/godaddy/asherah/pull/286
Phase 2: https://github.com/godaddy/asherah/pull/300

**Comment by jgowdy** [comment]
> the protections offered by SecureMemory aren't super-applicable to a managed shared execution environment like AWS Lambda.

I would respectfully assert that the protections offered by SecureMemory are only more relevant in a shared execution environment like AWS Lambda.  :)

**Comment by lyricnz** [comment]
Yup, I saw the work on the suballocator.  I'm not actually on the project anymore, but I let them know there could be a solution coming.  Thanks for your efforts.

**Comment by trulyshelton** [comment]
What is the latest on this? 
We are seeing this issue in both Lambda and ECS fargate. While we try to configure ulimit `memlock -1:-1` for fargate it doesn't seems to be working.

**Comment by lyricnz** [comment]
Seems that @sushantmimani has a new PR which might solve this #392 ?  Not merged yet

**Comment by sushantmimani** [comment]
> Seems that @sushantmimani has a new PR which might solve this #392 ? Not merged yet

@lyricnz  #392 is merged now and C# SecureMemory v0.2.0 is available for use. We'll release the updated AEL and Reference App shortly.

**Comment by lyricnz** [comment]
Excellent! Thank you!

**Comment by sushantmimani** [comment]
AEL v0.2.6 which uses the latest SecureMemory v0.2.6 is out now

**Comment by sushantmimani** [comment]
Closing due to no activity.

**Comment by lyricnz** [comment]
Thanks @sushantmimani .  The client has been using the code from #392 for about a year now, with good success.
